// handshake-login/build/main/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keyStringToArrayBuffer = exports.exportPrivateCryptoKey = exports.calculateFingerprint = exports.exportCryptoKey = exports.encodeText = exports.arrayBufferToString = exports.stringToArrayBuffer = exports.hash = exports.concatTypedArrays = exports.parseTxtRecordData = exports.decodeBase64 = exports.encodeBase64 = exports.SubtleCrypto = exports.Crypto = void 0;
if (typeof window !== 'undefined' && window.crypto) {
    exports.Crypto = window.crypto;
    exports.SubtleCrypto = exports.Crypto.subtle;
}
else {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    exports.Crypto = require('crypto').webcrypto;
    exports.SubtleCrypto = exports.Crypto.subtle;
}
// https://stackoverflow.com/a/62667703/1724828
/**
 * Encode a string of text as base64
 *
 * @param data The string of text.
 * @returns The base64 encoded string.
 */
function encodeBase64(data) {
    if (typeof btoa === 'function') {
        return btoa(data);
    }
    else if (typeof Buffer === 'function') {
        return Buffer.from(data, 'utf-8').toString('base64');
    }
    else {
        throw new Error('Failed to determine the platform specific encoder');
    }
}
exports.encodeBase64 = encodeBase64;
/**
 * Decode a string of base64 as text
 *
 * @param data The string of base64 encoded text
 * @returns The decoded text.
 */
function decodeBase64(data) {
    if (typeof atob === 'function') {
        return atob(data);
    }
    else if (typeof Buffer === 'function') {
        return Buffer.from(data, 'base64').toString('utf-8');
    }
    else {
        throw new Error('Failed to determine the platform specific decoder');
    }
}
exports.decodeBase64 = decodeBase64;
function parseTxtRecordData(str) {
    if (str.startsWith('"'))
        str = str.slice(1);
    if (str.endsWith('"'))
        str = str.slice(0, -1);
    const txtDataObj = str.split(';').reduce((data, part) => {
        const tmp = part.split('=');
        return Object.assign(Object.assign({}, data), { [tmp[0]]: tmp[1] });
    }, {});
    return txtDataObj;
}
exports.parseTxtRecordData = parseTxtRecordData;
function concatTypedArrays(arrays) {
    let length = 0;
    arrays.forEach((item) => {
        length += item.length;
    });
    // Create a new array with total length and merge all source arrays.
    const mergedArray = new Uint8Array(length);
    let offset = 0;
    arrays.forEach((item) => {
        mergedArray.set(item, offset);
        offset += item.length;
    });
    // Should print an array with length 90788 (5x 16384 + 8868 your source arrays)
    return mergedArray;
}
exports.concatTypedArrays = concatTypedArrays;
async function hash(text) {
    const textEncoder = new TextEncoder();
    const fingerprint = await exports.SubtleCrypto.digest('SHA-256', textEncoder.encode(text));
    const hashArray = Array.from(new Uint8Array(fingerprint));
    const hashHex = hashArray
        .map((b) => b.toString(16).padStart(2, '0'))
        .join(''); // convert bytes to hex string
    return hashHex;
}
exports.hash = hash;
function stringToArrayBuffer(str) {
    const buf = new ArrayBuffer(str.length);
    const bufView = new Uint8Array(buf);
    for (let i = 0, strLen = str.length; i < strLen; i++) {
        bufView[i] = str.charCodeAt(i);
    }
    return buf;
}
exports.stringToArrayBuffer = stringToArrayBuffer;
function arrayBufferToString(buf) {
    return String.fromCharCode.apply(null, Array.from(new Uint8Array(buf)));
}
exports.arrayBufferToString = arrayBufferToString;
const textEncoder = new TextEncoder();
function encodeText(str) {
    return textEncoder.encode(str);
}
exports.encodeText = encodeText;
async function exportCryptoKey(key) {
    const exported = await exports.SubtleCrypto.exportKey('spki', key);
    const exportedAsString = arrayBufferToString(exported);
    const exportedAsBase64 = encodeBase64(exportedAsString);
    return `-----BEGIN PUBLIC KEY-----\n${exportedAsBase64}\n-----END PUBLIC KEY-----`;
}
exports.exportCryptoKey = exportCryptoKey;
async function calculateFingerprint(key) {
    const exportedKey = await exportCryptoKey(key);
    const fingerprint = await exports.SubtleCrypto.digest('SHA-256', encodeText(exportedKey));
    const hashArray = Array.from(new Uint8Array(fingerprint));
    const hashHex = hashArray
        .map((b) => b.toString(16).padStart(2, '0'))
        .join(''); // convert bytes to hex string
    return hashHex;
}
exports.calculateFingerprint = calculateFingerprint;
async function exportPrivateCryptoKey(key) {
    const exported = await exports.SubtleCrypto.exportKey('pkcs8', key);
    const exportedAsString = arrayBufferToString(exported);
    const exportedAsBase64 = encodeBase64(exportedAsString);
    return `-----BEGIN PRIVATE KEY-----\n${exportedAsBase64}\n-----END PRIVATE KEY-----`;
}
exports.exportPrivateCryptoKey = exportPrivateCryptoKey;
function keyStringToArrayBuffer(pem) {
    const cleanKey = pem
        .replace(/\n?-{5}\n?[\w\s]+\n?-{5}\n?/g, '')
        .replace(/=+$/g, '');
    // base64 decode the string to get the binary data
    const binaryDerString = decodeBase64(cleanKey);
    // convert from a binary string to an ArrayBuffer
    return stringToArrayBuffer(binaryDerString);
}
exports.keyStringToArrayBuffer = keyStringToArrayBuffer;
