// handshake-login/build/main/lib/strategies/localStorage/cryptography.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifySignature = exports.sign = exports.importCryptoPrivateKey = exports.importCryptoKey = exports.generateKeyPair = void 0;
const utils_1 = require("../../utils");
async function generateKeyPair() {
    return utils_1.SubtleCrypto.generateKey({
        name: 'RSA-PSS',
        modulusLength: 4096,
        publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
        hash: 'SHA-512',
    }, true, ['sign', 'verify']);
}
exports.generateKeyPair = generateKeyPair;
function importCryptoKey(pem) {
    const binaryDer = utils_1.keyStringToArrayBuffer(pem);
    return utils_1.SubtleCrypto.importKey('spki', binaryDer, {
        name: 'RSA-PSS',
        hash: 'SHA-512',
    }, true, ['verify']);
}
exports.importCryptoKey = importCryptoKey;
function importCryptoPrivateKey(pem) {
    const binaryDer = utils_1.keyStringToArrayBuffer(pem);
    return utils_1.SubtleCrypto.importKey('pkcs8', binaryDer, {
        name: 'RSA-PSS',
        hash: 'SHA-512',
    }, true, ['sign']);
}
exports.importCryptoPrivateKey = importCryptoPrivateKey;
async function sign(privateKey, data) {
    const signature = await utils_1.SubtleCrypto.sign({
        name: 'RSA-PSS',
        saltLength: 64,
    }, privateKey, utils_1.encodeText(data));
    const exportedAsString = utils_1.arrayBufferToString(signature);
    const exportedAsBase64 = utils_1.encodeBase64(exportedAsString);
    return exportedAsBase64;
}
exports.sign = sign;
async function verifySignature(publicKey, signature, data) {
    const binaryDerString = utils_1.decodeBase64(signature);
    const binaryDer = utils_1.stringToArrayBuffer(binaryDerString);
    return await utils_1.SubtleCrypto.verify({
        name: 'RSA-PSS',
        saltLength: 64,
    }, publicKey, binaryDer, utils_1.encodeText(data));
}
exports.verifySignature = verifySignature;
