// handshake-login/build/main/lib/strategies/localStorage/localStorage.js
"use strict";
// https://github.com/namebasehq/handshake-id-manager/blob/master/src/services/DeviceService/index.ts
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalStorageStrategy = void 0;
const utils_1 = require("../../utils");
const abstract_1 = require("../abstract");
const cryptography = __importStar(require("./cryptography"));
const store = __importStar(require("./store"));
/**
 * Local Storage Strategy
 */
class LocalStorageStrategy extends abstract_1.AbstractStrategy {
    constructor() {
        super();
        this.strategyName = 'LocalStorageStrategy';
        this.storageKey = 'handshake:login:deviceId';
    }
    /**
     * Reset all properties
     */
    reset() {
        this.domain = null;
        this.identity = null;
        this.reqData = null;
        this.resData = null;
        this.deviceRecord = null;
    }
    /**
     * Set Request Data
     * @param reqData Request data to set
     */
    setRequestData(reqData) {
        this.reset();
        this.reqData = reqData;
        this.domain = reqData.domain.toLowerCase();
    }
    /**
     * Set Device Record
     * @param deviceRecord Device record to set
     */
    setDeviceRecord(deviceRecord) {
        this.deviceRecord = deviceRecord;
    }
    /**
     * Set Response Data
     * @param resData Response data to set
     */
    setResponseData(resData) {
        this.resData = resData;
        this.domain = resData.domain;
    }
    /**
     * Get Device ID
     * and generate if it doesn't exist
     * @returns device id
     */
    async getDeviceId() {
        var _a;
        return ((_a = window.localStorage.getItem(this.storageKey)) !== null && _a !== void 0 ? _a : this._generateNewDeviceId());
    }
    /**
     * Generate and store new Device ID
     * @returns new device id
     * @private
     */
    _generateNewDeviceId() {
        const uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            const r = (Math.random() * 16) | 0, v = c == 'x' ? r : (r & 0x3) | 0x8;
            return v.toString(16);
        });
        window.localStorage.setItem(this.storageKey, uuid);
        return uuid;
    }
    /**
     * Generate new Identity
     * @returns Identity
     */
    async generateIdentity() {
        var _a;
        if (!this.domain)
            throw new Error('Domain not set.');
        const { privateKey, publicKey } = await cryptography.generateKeyPair();
        const publicKeyString = await utils_1.exportCryptoKey(publicKey);
        const identity = {
            name: this.domain,
            publicKey: publicKeyString,
            privateKey: await utils_1.exportPrivateCryptoKey(privateKey),
            fingerprint: await utils_1.hash(publicKeyString),
        };
        await store.saveIdentity(identity, (_a = this.strategyOptions) === null || _a === void 0 ? void 0 : _a.password);
        this.identity = identity;
        return identity;
    }
    /**
     * Get an Identity
     * @returns Identity
     */
    async getIdentity() {
        var _a;
        if (!this.domain)
            throw new Error('Domain not set.');
        this.identity = await store.getIdentity(this.domain, (_a = this.strategyOptions) === null || _a === void 0 ? void 0 : _a.password);
        return this.identity;
    }
    /**
     * Sign challenge
     * @returns signed data
     */
    async sign() {
        var _a, _b;
        if (!this.domain)
            throw new Error('Domain not set.');
        if (!((_a = this.reqData) === null || _a === void 0 ? void 0 : _a.challenge))
            throw new Error('Challenge not set.');
        if (!((_b = this.identity) === null || _b === void 0 ? void 0 : _b.privateKey))
            throw new Error('Private Key not set.');
        let privateKey;
        try {
            privateKey = await cryptography.importCryptoPrivateKey(this.identity.privateKey);
        }
        catch (error) {
            console.error(error);
            throw new Error('Could not import invalid private key.');
        }
        const signature = await cryptography.sign(privateKey, this.reqData.challenge);
        return utils_1.encodeBase64(signature);
    }
    /**
     * Verify signature
     * @param challenge challenge to verify signature with
     * @returns boolean if signature verified
     */
    async verify(challenge) {
        var _a, _b;
        if (!challenge)
            throw new Error('Challenge not set.');
        if (!this.domain)
            throw new Error('Domain not set.');
        if (!((_a = this.resData) === null || _a === void 0 ? void 0 : _a.publicKey))
            throw new Error('Public Key not set.');
        if (!((_b = this.resData) === null || _b === void 0 ? void 0 : _b.signed))
            throw new Error('Signature not set.');
        let publicKey;
        try {
            publicKey = await cryptography.importCryptoKey(this.resData.publicKey);
        }
        catch (error) {
            console.error(error);
            throw new Error('Could not import invalid public key.');
        }
        const signatureVerified = await cryptography.verifySignature(publicKey, this.resData.signed, challenge);
        return signatureVerified;
    }
    /**
     * Generate DNS Record for Identity
     * @param prefix derived from domain and device id
     * @returns DNSRecord
     */
    async generateDnsRecord(prefix) {
        var _a;
        if (!prefix)
            throw new Error('Prefix not set.');
        if (!this.domain)
            throw new Error('Domain not set.');
        if (!((_a = this.identity) === null || _a === void 0 ? void 0 : _a.fingerprint))
            throw new Error('Fingerprint not set.');
        return {
            type: 'TXT',
            name: `${prefix}._auth.${this.domain}.`,
            value: `v=0;fingerprint=${this.identity.fingerprint}`,
        };
    }
    /**
     * Generate Signature Data
     * @returns partial response data
     */
    async generateSignatureData() {
        return {
            signed: await this.sign(), // Buffer => b64 => b64 (https://github.com/namebasehq/handshake-id-manager/issues/6)
        };
    }
    /**
     * Get Fingerprint from Response Data's Public Key
     * @returns fingerprint
     */
    async getFingerprint() {
        var _a;
        if (!((_a = this.resData) === null || _a === void 0 ? void 0 : _a.publicKey))
            throw new Error('Public Key not set.');
        let pubKey;
        try {
            pubKey = await cryptography.importCryptoKey(this.resData.publicKey);
        }
        catch (error) {
            console.error(error);
            throw new Error('Could not import invalid public key.');
        }
        return utils_1.calculateFingerprint(pubKey);
    }
}
exports.LocalStorageStrategy = LocalStorageStrategy;
