// handshake-login/build/main/lib/strategies/localStorage/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveIdentity = exports.getIdentity = void 0;
const utils_1 = require("../../utils");
const identityKeyPrefix = 'handshake:login:identity_';
const ivKeyPrefix = 'handshake:login:iv_';
async function getIdentity(domain, password) {
    if (!domain)
        throw new Error('Domain not set.');
    const encryptedIdentity = window.localStorage.getItem(`${identityKeyPrefix}${utils_1.encodeBase64(domain)}`);
    if (!encryptedIdentity)
        return null;
    const iv = utils_1.stringToArrayBuffer(window.localStorage.getItem(`${ivKeyPrefix}${utils_1.encodeBase64(domain)}`));
    try {
        const identity = await decrypt(utils_1.stringToArrayBuffer(utils_1.decodeBase64(encryptedIdentity)), await getPrivateKey(password), iv);
        return JSON.parse(utils_1.arrayBufferToString(identity));
    }
    catch (error) {
        throw new Error('Invalid password.');
    }
}
exports.getIdentity = getIdentity;
async function saveIdentity(identity, password) {
    if (!identity)
        throw new Error('Identity not set.');
    const iv = utils_1.Crypto.getRandomValues(new Uint8Array(12));
    window.localStorage.setItem(`${ivKeyPrefix}${utils_1.encodeBase64(identity.name)}`, utils_1.arrayBufferToString(iv));
    const encryptedIdentity = await encrypt(utils_1.stringToArrayBuffer(JSON.stringify(identity)), await getPrivateKey(password), iv);
    window.localStorage.setItem(`${identityKeyPrefix}${utils_1.encodeBase64(identity.name)}`, utils_1.encodeBase64(utils_1.arrayBufferToString(encryptedIdentity)));
}
exports.saveIdentity = saveIdentity;
async function getPrivateKey(password) {
    const k = (await utils_1.hash(await utils_1.hash(password))).slice(0, 22);
    const privateKey = await utils_1.SubtleCrypto.importKey('jwk', {
        key_ops: ['encrypt', 'decrypt'],
        ext: true,
        kty: 'oct',
        k: k,
        alg: 'A128GCM',
    }, {
        name: 'AES-GCM',
    }, false, ['encrypt', 'decrypt']);
    return privateKey;
}
async function encrypt(data, privateKey, iv) {
    if (!data)
        throw new Error('Data to encrypt not set.');
    if (!privateKey)
        throw new Error('Private Key not set.');
    if (!iv)
        throw new Error('IV not set.');
    const encryptedData = await utils_1.SubtleCrypto.encrypt({
        name: 'AES-GCM',
        iv: iv,
    }, privateKey, data);
    return encryptedData;
}
async function decrypt(data, privateKey, iv) {
    if (!data)
        throw new Error('Data to decrypt not set.');
    if (!privateKey)
        throw new Error('Private Key not set.');
    if (!iv)
        throw new Error('IV not set.');
    const decryptedData = await utils_1.SubtleCrypto.decrypt({
        name: 'AES-GCM',
        iv: iv,
    }, privateKey, data);
    return decryptedData;
}
