// handshake-login/build/module/lib/strategies/localStorage/cryptography.js
import { arrayBufferToString, decodeBase64, encodeBase64, encodeText, keyStringToArrayBuffer, stringToArrayBuffer, SubtleCrypto, } from '../../utils';
export async function generateKeyPair() {
    return SubtleCrypto.generateKey({
        name: 'RSA-PSS',
        modulusLength: 4096,
        publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
        hash: 'SHA-512',
    }, true, ['sign', 'verify']);
}
export function importCryptoKey(pem) {
    const binaryDer = keyStringToArrayBuffer(pem);
    return SubtleCrypto.importKey('spki', binaryDer, {
        name: 'RSA-PSS',
        hash: 'SHA-512',
    }, true, ['verify']);
}
export function importCryptoPrivateKey(pem) {
    const binaryDer = keyStringToArrayBuffer(pem);
    return SubtleCrypto.importKey('pkcs8', binaryDer, {
        name: 'RSA-PSS',
        hash: 'SHA-512',
    }, true, ['sign']);
}
export async function sign(privateKey, data) {
    const signature = await SubtleCrypto.sign({
        name: 'RSA-PSS',
        saltLength: 64,
    }, privateKey, encodeText(data));
    const exportedAsString = arrayBufferToString(signature);
    const exportedAsBase64 = encodeBase64(exportedAsString);
    return exportedAsBase64;
}
export async function verifySignature(publicKey, signature, data) {
    const binaryDerString = decodeBase64(signature);
    const binaryDer = stringToArrayBuffer(binaryDerString);
    return await SubtleCrypto.verify({
        name: 'RSA-PSS',
        saltLength: 64,
    }, publicKey, binaryDer, encodeText(data));
}
